File: frontend/src/types/notifications.ts
/**
 * 通知展示投影（PRD §2.6 / design §5.4）：kind/domain_status 的中文标签与
 * 三分区归类纯函数。只做展示层分类，不改变任何领域状态，也不推导授权。
 *
 * 三个分区严格按服务端字段归类：
 * - 待我处理：ActionCard 引用且领域仍 pending（去处理走既有 ActionCard 流程）；
 * - 已完成·历史：read_at 有值且领域已到终态（非 pending 即视为该通知的稳定态）；
 * - 通知：其余（含未读的领域终态告知、bridge/relation/membership 通知）。
 */
import type { NotificationDomainStatus, NotificationItem, NotificationKind } from '@/types/api'

export type NotificationSection = 'pending-action' | 'notices' | 'history'

export const NOTIFICATION_KIND_LABELS: Record<NotificationKind, string> = {
  action_card: '管家建议',
  space_membership: '空间成员',
  bridge: '家族连接',
  relation: '关系',
}

export const NOTIFICATION_DOMAIN_STATUS_LABELS: Record<NotificationDomainStatus, string> = {
  pending: '待处理',
  active: '已生效',
  accepted: '已接受',
  rejected: '已拒绝',
  cancelled: '已取消',
  revoked: '已撤销',
  expired: '已过期',
  withdrawn: '已撤回',
  removed: '已移除',
  done: '已完成',
}

/** 领域状态徽章阶（design.md §3.4 tokens.css 工具类，双主题同源） */
export const NOTIFICATION_DOMAIN_STATUS_BADGES: Record<NotificationDomainStatus, string> = {
  pending: 'fg-badge--accent',
  active: 'fg-badge--confirmed',
  accepted: 'fg-badge--confirmed',
  rejected: 'fg-badge--disputed',
  cancelled: 'fg-badge--neutral',
  revoked: 'fg-badge--disputed',
  expired: 'fg-badge--provisional',
  withdrawn: 'fg-badge--neutral',
  removed: 'fg-badge--neutral',
  done: 'fg-badge--confirmed',
}

/** 通知分区归类：ActionCard 引用且未处理 → 待我处理；已读且领域终态 → 历史 */
export function classifyNotification(item: NotificationItem): NotificationSection {
  if (item.kind === 'action_card' && item.domain_status === 'pending') {
    return 'pending-action'
  }
  if (item.read_at !== null && item.domain_status !== 'pending') {
    return 'history'
  }
  return 'notices'
}

export function classifyNotifications(items: readonly NotificationItem[]): {
  pendingAction: NotificationItem[]
  notices: NotificationItem[]
  history: NotificationItem[]
} {
  const pendingAction: NotificationItem[] = []
  const notices: NotificationItem[] = []
  const history: NotificationItem[] = []
  for (const item of items) {
    const section = classifyNotification(item)
    if (section === 'pending-action') pendingAction.push(item)
    else if (section === 'history') history.push(item)
    else notices.push(item)
  }
  return { pendingAction, notices, history }
}

File: backend/app/schemas/notifications.py
"""Notifications API contract（字段白名单，与前端 decoder 严格对齐）。

payload 字段允许明文字符串、``{"__masked__": true}`` 哨兵或 null；
domain_status 是跨领域 FSM 的最小状态集，ActionCard 引用只含 card_id/revision。
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

NotificationKind = Literal[
    "action_card", "space_membership", "bridge", "relation", "steward_suggestion"
]
NotificationDomainStatus = Literal[
    "pending",
    "active",
    "accepted",
    "rejected",
    "cancelled",
    "revoked",
    "expired",
    "withdrawn",
    "removed",
    "done",
]
MaskableString = str | dict[str, Any] | None


class NotificationPayloadOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str
    summary: MaskableString
    actor_name: MaskableString
    space_name: MaskableString


class NotificationActionCardRefOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    card_id: int
    revision: int


class NotificationSuggestionRefOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    suggestion_id: int


class NotificationItemOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: int
    space_id: int
    kind: NotificationKind
    payload: NotificationPayloadOut
    domain_status: NotificationDomainStatus
    action_card: NotificationActionCardRefOut | None
    suggestion: NotificationSuggestionRefOut | None
    created_at: datetime
    read_at: datetime | None


class NotificationsPageOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    space_id: int
    items: list[NotificationItemOut]
    unread_count: int


class NotificationReadAllIn(BaseModel):
    model_config = ConfigDict(extra="forbid")

    space_id: int = Field(gt=0)


class NotificationReadOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: int
    read_at: datetime


class NotificationReadAllOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    space_id: int
    marked_count: int


__all__ = [
    "NotificationActionCardRefOut",
    "NotificationDomainStatus",
    "NotificationItemOut",
    "NotificationKind",
    "NotificationReadAllIn",
    "NotificationReadAllOut",
    "NotificationReadOut",
    "NotificationSuggestionRefOut",
    "NotificationsPageOut",
    "NotificationPayloadOut",
]
